### PR TITLE
HelixBootstrapTool changes to support InstanceConfig-To-PropertyStore migration

### PR DESCRIPTION
**Pull request by jsjtzyy**
Ambry will eventually migrate datanode info (i.e. disk, replica) from InstanceConfig to PropertyStore in Helix. This PR makes several changes in HelixBootstrapUpgradeTool to support both. User should be able to specify which DataNodeConfigSource type to use and the tool will update Helix accordingly.

### CODE REVIEW

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java [hunk 1]
```diff
@@ -929,90 +942,185 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
...
+    DataNodeConfig dataNodeConfig;
+    if (dataNodeConfigSourceType == DataNodeConfigSourceType.PROPERTY_STORE) {
+      dataNodeConfig = adapter.get(instanceName);
+    } else {
+      dataNodeConfig = converter.convert(adminForDc.get(dcName).getInstanceConfig(clusterName, instanceName));
+    }
+    return dataNodeConfig;
+  }
+
+  private boolean areDataNodeConfigsEquivalent(DataNodeConfig configInStatic, DataNodeConfig configInHelix,
```
Comment (cgtz) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1068]: Maybe we could move this method into `DataNodeConfig`

`public boolean equals(DataNodeConfig other, boolean ignorePartitionStateConfigs)` (or whatever name you feel is best for the sealed/stopped/disabled fields)

**Review by cgtz — APPROVED** [review]
Thank you for making this tool improvement! Mostly LGTM

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java [hunk 2]
```diff
@@ -366,7 +366,7 @@ private boolean addNewReplicaInfo(ReplicaId replicaId) {
       newReplicaInfoAdded = true;
     }
     if (newReplicaInfoAdded) {
-      logger.info("Updating config: {} in Helix by adding partition {}", dataNodeConfig, partitionName);
+      //logger.info("Updating config: {} in Helix by adding partition {}", dataNodeConfig, partitionName);
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixParticipant.java:369]: Sorry, I commented it for debugging. Will remove it.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java [hunk 3]
```diff
@@ -88,6 +88,22 @@ public void close() {
     propertyStore.stop();
   }
 
+  /**
+   * Remove data config from property store
+   * @param instanceName the name of instance
+   */
+  public void remove(String instanceName) {
+    propertyStore.remove(CONFIG_PATH + "/" + instanceName, AccessOption.PERSISTENT);
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java:96]: Sure, will do.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/DataNodeConfig.java [hunk 4]
```diff
@@ -37,7 +38,7 @@
   private final Set<String> sealedReplicas = new HashSet<>();
   private final Set<String> stoppedReplicas = new HashSet<>();
   private final Set<String> disabledReplicas = new HashSet<>();
-  private final Map<String, DiskConfig> diskConfigs = new HashMap<>();
+  private final Map<String, DiskConfig> diskConfigs = new TreeMap<>();
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/DataNodeConfig.java:41]: Not really for readability. The Helix bootstrap tool sometimes reported the instanceConfig was updated on existing node (however, we didn't touch those existing nodes). The DiskConfigs in Helix and static were actually same except for ordering. This change should be able suppress such messages (making the BootstrapTool less confusing).

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java (hunk 1 above)
Comment (jsjtzyy) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java]: Sure, it makes more sense.

**Review by cgtz — APPROVED** [review]

**Review by zzmao — APPROVED** [review]
LGTM

**Pull request merged**